### ISSUE TITLE
Voxel Editors and Functional Modules?

### ISSUE THREAD

**Issue by MajorLunaC** [konstructs/client#104]
I'm not sure if this is appropriate and all, but I was wondering if some unique attracting features could be introduced so that this wont just be minecraft clone #358 (fake number). Of course, great look and great performance come first and foremost, practically as a bare minimum requirement nowadays. But ultimately, if you want this to attract any attention, this game needs to stand out significantly. The top features I've seen in demand (even in minecraft itself) are features themselves, and customization (including easy modding).

1) Features: For starters, liquids, but best of all and in highest demand would be custom vehicles. That may sound overwhelming, and some aspects might be, but on a basic level, it could function fairly easily. It would consist of pre-programmed functional modules, which you can think of like little invisible boats that interact with things, to which you can attach as many blocks as you like. Same with every other type of vehicle one could think of.

Of course it could work on a physics basis instead (or also), where you add buoyant blocks to keep the structure floating.

The biggest challenge would be the platform-like physics, where you can walk on the boat while it moves. Of course, you could instead have built mount points (pilot seat, cannons 1-10, deck, under deck, etc.) placed by the client before attaching the functional module. This means you could select and switch through seats that players would be stuck to, with a set range of motion defined by the builder when placing them.

2) Customization with Voxel Editors: I've noticed that some block-like games have certain formats that allow design using small voxels, much like the smaller parts of character models and animals in minecraft. There are already several voxel editors for various games and formats:

VoxelShop (Tool; Multi-Platform): https://blackflux.com/index.php
Zoxel (Tool; Multi-Platform): https://github.com/grking/zoxel
MagicaVoxel (Tool; Window$ + Mac): https://voxel.codeplex.com/
Slab6 (Tool; Window$ Only): http://aos.gamebanana.com/tools/download/5195

Having such a feature would allow people to design their own items as they want them, and really virtually anything in the game, including any special block (maybe even liquids?). These could be server-based mods, from what I understand, based on the present design, although I wonder if clients could bring some with them too, as long as the one running the server gets a look at them and approves them. These too could have pre-made functional modules, although mostly just ones that attach to other blocks.

In general there should be easy ways to mod, as they could sustain a game indefinitely, as long as patches don't destroy them (number 1 cause of mod and game death, number 2 being jerks that can't be dealt with).

Well, these are all theoretical recommendations, it's just that it would be best to consider them early as it might be harder to allow for them farther down the road. At the very least consider making a rollercoaster system where you could place and travel along a rail on any surface of a block, even upside-down. It is certainly possible if the minecraft-style rules are applied from every direction.

I hope this gives you guys something to think about. Thanks for any consideration, and thanks for reading! :)

**Comment by nsg** [comment]
> I was wondering if some unique attracting features could be introduced so that this wont just be minecraft clone #358 (fake number). /../ But ultimately, if you want this to attract any attention, this game needs to stand out significantly.

Agree, we like to difference our a little from classic Minecraft, no point to build the same game again. At the moment things are similar on the surface but we are working on that :)

> 1) Features: For starters, liquids, but best of all and in highest demand would be custom vehicles.

Exactly what we are planning to do, a little inspiration from Forge based Mincraft mods like Redpower frames: https://youtu.be/KJZgsk4pCB0?t=1m31s We intend smooths 360 degrees unlimited movements. We have actually already spent a lot of time with the graphic engine to make this fast and possible.

At the moment I think that the most important thing is to make the game playable, walk, swim, mine, build, craft and a interesting world to play in. After that, things like this is a possibility :)

> 2) Customization with Voxel Editors

The biggest change is to keep the size of the block format down, remember that we are processing a lot more blocks compared to Minecraft, look at this demo https://www.youtube.com/watch?v=GiJIvIH6FVw ... everything needs to be processed by the server, sent over the network, processed on your computer and uploaded to the graphic card. So even a tiny bit of extra work is noticed when you multiply it with millions or even billions.

At the moment blocks can only have two shapes, a block and a cross (flowers and so). We are thinking about increasing this to 7 more states to get something similar to microblocks: https://youtu.be/YYvuOeFppk8?t=23s

It _should_ be possible to write a simple plugin that can import custom constructions from Voxel Editors. At the moment we have no plans to allow players to "world edit" the world and upload models but everything is really pluggable and maybe someone else will do so. Noting is written in stone so you never know :)

> Well, these are all theoretical recommendations, it's just that it would be best to consider them early as it might be harder to allow for them farther down the road.

Indeed, ideas are always welcome. I'm sure the game will be quite unique when it's in a playable state next year. After that we will work hard to add features and fun stuff to the game, possible things you suggested :)

> At the very least consider making a rollercoaster system where you could place and travel along a rail on any surface of a block, even upside-down.

We will most likely not add things like Minecraft styled boats and minecarts, we like to keep the game in the voxel blocky style where you have to build your boats and crafts, what you suggested under `1)`.

**Comment by DanielOaks** [comment]
This is a little unrelated, but I was adding a lua-based module system to my Craft fork a while ago. I didn't get too far with it or get it to the point I'd like, but if it looks useful feel free to scavenge anything from it for your client: [api.c](https://github.com/DanielOaks/Craft/blob/dynamic_stuff/src/api.c) [api.h](https://github.com/DanielOaks/Craft/blob/dynamic_stuff/src/api.h) [water.lua](https://github.com/DanielOaks/Craft/blob/dynamic_stuff/modules/water.lua)

**Comment by nsg** [comment]
@DanielOaks 

Oh, that's cool. We will have a look at it.

Because most things happens in the server at the moment we will most likely not add extendibility to the client at this time. I can see a future where we use lua to push out more complex behaviors from the server, or, using lua to customize the client.

Thank you, I will remember your fork for the future.

**Comment by MajorLunaC** [comment]
That sounds really promising! I've been trying to learn C in the little spare time I have, but I doubt I could help much.

As for the Lua scripting part, although you could easily attract people from much more popular minecraft-like games like Minetest and all 3+ of its derivations, performance has been an ever-present issue there, especially online. Maybe one of the scripting languages that compiles to C/C++ or close to it. Even Clojure, and other Lisp-based languages have compilers to C.
